src/core/processing/qgsprocessingparametermeshdataset.cpp

```diff
@@ -0,0 +1,201 @@
+{
+  Q_UNUSED( context );
+  if ( !input.isValid() )
+    return false;
+
+  if ( input.type() != QVariant::List )
+    return false;
+  const QVariantList list = input.toList();
+  return list.count() > 0;
+}
+
+QString QgsProcessingParameterMeshDatasetGroups::valueAsPythonString( const QVariant &value, QgsProcessingContext &context ) const
+{
+  Q_UNUSED( context );
+  QStringList parts;
+  const QVariantList variantDatasetGroupIndexes = value.toList();
+  for ( const QVariant &variantIndex : variantDatasetGroupIndexes )
+    parts.append( QString::number( variantIndex.toInt() ) );
+
+  return parts.join( ',' ).prepend( '[' ).append( ']' );
+}
+
+QString QgsProcessingParameterMeshDatasetGroups::asPythonString( QgsProcessing::PythonOutputType outputType ) const
+{
+  switch ( outputType )
+  {
+    case QgsProcessing::PythonQgsProcessingAlgorithmSubclass:
+    {
+      QString code = QStringLiteral( "QgsProcessingParameterMeshDatasetGroup('%1', '%2')" ).arg( name(), description() );
+      return code;
+    }
+  }
+  return QString();
+}
+
+QStringList QgsProcessingParameterMeshDatasetGroups::dependsOnOtherParameters() const
+{
+  return QStringList() << mMeshLayerParameterName;
+}
+
+QString QgsProcessingParameterMeshDatasetGroups::meshLayerParameterName() const
+{
+  return mMeshLayerParameterName;
+}
+
+QgsProcessingParameterMeshDatasetTime::QgsProcessingParameterMeshDatasetTime(
+  const QString &name,
+  const QString &description,
+  const QString &meshLayerParameterName,
+  const QString &datasetGroupParameterName,
+  bool optional )
+  : QgsProcessingParameterDefinition( name, description, QVariant(), optional )
+  , mMeshLayerParameterName( meshLayerParameterName )
+  , mDatasetGroupParameterName( datasetGroupParameterName )
+{
+
+}
+
+QgsProcessingParameterDefinition *QgsProcessingParameterMeshDatasetTime::clone() const
+{
+  return new QgsProcessingParameterMeshDatasetTime( name(), description(), mMeshLayerParameterName, mDatasetGroupParameterName );
+}
+
+QString QgsProcessingParameterMeshDatasetTime::type() const
+{
+  return typeName();
+}
+
+bool QgsProcessingParameterMeshDatasetTime::checkValueIsAcceptable( const QVariant &input, QgsProcessingContext *context ) const
+{
+  Q_UNUSED( context );
+  if ( !input.isValid() )
+    return false;
+
+  if ( input.type() != QVariant::Map )
+    return false;
+  const QVariantMap map = input.toMap();
+  if ( ! map.contains( QStringLiteral( "type" ) ) )
+    return false;
+
+  QString type = map.value( QStringLiteral( "type" ) ).toString();
+  QVariant value = map.value( QStringLiteral( "value" ) );
+
+  if ( type == QStringLiteral( "static" ) )
+    return true;
+
+  if ( type == QStringLiteral( "dataset-time-step" ) )
+  {
+    if ( value.type() != QVariant::List )
+      return false;
+    if ( value.toList().count() != 2 )
+      return false;
+  }
+  else if ( type == QStringLiteral( "current-canvas-time" ) || type == QStringLiteral( "defined-date-time" ) )
```

Comment:
it is set here : https://github.com/qgis/QGIS/blob/48be634a43bdd4f790d962c71a37800c8fe90665/src/gui/processing/qgsprocessingmeshdatasetwidget.cpp#L560
I was thinking it was enough ... this option is set only if there is a map canvas in the `QgsProcessingParameterWidgetContext`.

Reply:
> But maybe in the QgsProcessingParameter no need to have this option available. Just consider this option as date/time defined. The choice is only done in the widget.

Yeah - there's confusion here because the implementation looks like it should accept a value of "current-canvas-time", but in fact that has no meaning outside of the widget.

I'd still be inclined to push for adding the time range to the context. I think that would be a nice flexible approach which will pay off for us later. It also means that an execution using "current-context-time" will work correctly when re-run from the processing history dialog, because it would correctly reflect the current canvas time instead of the time at the point of last execution...

Reply:
> You'd also need to add a way to set the context time via the qgis_process standalone tool, just like how the context's area/distance units can be set.
Maybe a string with ISO8601 format could be entered after --CURRENT_TIME. and we could have a function like this : https://github.com/lutraconsulting/MDAL/blob/f65103aab2df2ba7676ffadd0b98caef0a75cbef/mdal/mdal_datetime.hpp#L71

Any idea where this function could be implemented in QGIS code ?